core/git.go

```diff
@@ -254,92 +463,240 @@ func (ref *RemoteGitRef) PBDefinitions(ctx context.Context) ([]*pb.Definition, e
 	return nil, nil
 }
 
-func (ref *RemoteGitRef) Tree(ctx context.Context, srv *dagql.Server, discardGitDir bool) (*Directory, error) {
-	st, err := ref.getState(ctx, discardGitDir)
-	if err != nil {
-		return nil, err
+func (ref *RemoteGitRef) Tree(ctx context.Context, srv *dagql.Server, discardGitDir bool, depth int) (_ *Directory, rerr error) {
+	op, ok := DagOpFromContext[FSDagOp](ctx)
+	if !ok {
+		return nil, fmt.Errorf("no dagop")
 	}
-	return NewDirectorySt(ctx, ref.Query, st, "/", ref.Repo.Platform, ref.Repo.Services)
-}
+	cacheKey := dagql.CurrentID(ctx).Digest().Encoded()
 
-func (ref *RemoteGitRef) Commit(ctx context.Context) (string, error) {
-	bk, err := ref.Query.Buildkit(ctx)
+	locker := ref.Query.Locker()
+	locker.Lock(gitSnapshotIndex + cacheKey)
+	defer locker.Unlock(gitSnapshotIndex + cacheKey)
+	sis, err := searchGitSnapshot(ctx, op.Cache(), cacheKey)
 	if err != nil {
-		return "", fmt.Errorf("failed to get buildkit client: %w", err)
+		return nil, fmt.Errorf("failed to search metadata for %s: %w", cacheKey, err)
 	}
-	st, err := ref.getState(ctx, true)
-	if err != nil {
-		return "", err
+	if len(sis) > 0 {
+		res, err := op.Cache().Get(ctx, sis[0].ID(), nil)
+		if err != nil {
+			return nil, err
+		}
+		checkout := NewDirectory(ref.Query, nil, "/", ref.Query.Platform(), nil)
+		checkout.Result = res
+		return checkout, nil
 	}
 
-	if len(ref.Repo.Services) > 0 {
-		svcs, err := ref.Query.Services(ctx)
+	var checkoutRef bkcache.MutableRef
+	defer func() {
+		if rerr != nil && checkoutRef != nil {
+			checkoutRef.Release(context.WithoutCancel(ctx))
+		}
+	}()
+
+	err = ref.Repo.mountRemote(ctx, op, func(remote string) error {
+		git, cleanup, err := ref.Repo.setup(ctx)
 		if err != nil {
-			return "", err
+			return err
 		}
-		detach, _, err := svcs.StartBindings(ctx, ref.Repo.Services)
+		defer cleanup()
+		git = git.New(gitutil.WithGitDir(remote))
+		gitDir, err := git.GitDir(ctx)
 		if err != nil {
-			return "", err
+			return fmt.Errorf("could not find git dir: %w", err)
+		}
+
+		// skip fetch if commit already exists
+		doFetch := true
+		if res, err := git.Run(ctx, "rev-parse", ref.FullRef); err == nil && string(res) == ref.Commit {
+			doFetch = false
 		}
-		defer detach()
+
+		if doFetch {
+			svcs, err := ref.Repo.Query.Services(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to get services: %w", err)
+			}
+			detach, _, err := svcs.StartBindings(ctx, ref.Repo.Services)
+			if err != nil {
+				return err
+			}
+			defer detach()
```

Comment:
Heh, pretty happy with this - cc @vito 

This lets us *only* spin up services when we're doing the operations that need them - if we don't do a `git fetch`, then we never need to access them.